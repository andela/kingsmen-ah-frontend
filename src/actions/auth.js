import axios from 'axios';
import { toast } from 'react-toastify';
import {
  SIGNIN_FAILURE,
  SIGNIN_SUCCESS,
  GET_PROFILE_ERROR,
  SET_CURRENT_USER,
  IS_LOADING,
  SET_PROFILE,
  LOGOUT_USER,
  RESET_PROFILE
} from './types';

axios.defaults.baseURL =
  'https://kingsmen-ah-backend-staging.herokuapp.com/api/v1';

export const setAuthToken = token => {
  if (token) {
    axios.defaults.headers.common['Authorization'] = `Bearer ${token}`;
  } else {
    delete axios.defaults.headers.common['Authorization'];
  }
};
export const setCurrentUser = user => {
  return {
    type: SET_CURRENT_USER,
    payload: user
  };
};

export const isLoading = value => ({
  type: IS_LOADING,
  payload: value
});

export const setUserProfile = payload => ({
  type: SET_PROFILE,
  payload
});

export const getProfile = username => async dispatch => {
  try {
    const res = await axios.get(`/profiles/${username}`);
    dispatch(setUserProfile(res.data.payload));
    window.location.reload();
  } catch (error) {
    if (error.response) {
      return dispatch({
        type: GET_PROFILE_ERROR,
        payload: error.response.data.errors
      });
    }
    toast.error('Please check your network connection and try again');
  }
};

export const logoutUser = history => dispatch => {
  dispatch(isLoading(true));
  axios.post('/auth/logout');
  localStorage.removeItem('jwtToken');
  dispatch({ type: LOGOUT_USER });
  dispatch({type: RESET_PROFILE })
  history.push('/');
  setAuthToken();
};

export const loginUser = userData => async dispatch => {
  try {
    dispatch(isLoading(true));

    const res = await axios.post('/auth/login', userData);
    const response = res.data.payload;

    const { token } = response;
    localStorage.setItem('jwtToken', token);

    const { id, email, username, exp } = response;
    const user = {
      id,
      email,
      username,
      exp
    };

    setAuthToken(token);
    dispatch(setCurrentUser(user));
    toast.success('Login successful');
    dispatch(getProfile(username));
    dispatch({ type: SIGNIN_SUCCESS });
  } catch (error) {
    if (error.response) {
      const errors = error.response.data.errors;
      if (errors.global) toast.error(errors.global);
      return dispatch({
        type: SIGNIN_FAILURE,
        payload: errors
      });
    }
    dispatch(isLoading(false));
    toast.error('Please check your network connection and try again');
  }
};
