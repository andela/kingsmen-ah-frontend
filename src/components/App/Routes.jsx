import React from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import Home from '@components/views/Home';
import Login from '@components/views/Login';
import NotFound from '@components/views/NotFound';

const Routes = () => (
  <Router>
    <Switch>
      <Route exact path="/" component={Home} />
      <Route exact path="/login" component={Login} />
      <Route path="/" component={NotFound} />
    </Switch>
  </Router>
);

export default Routes;
