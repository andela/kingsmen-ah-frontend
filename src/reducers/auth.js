import {
  SET_CURRENT_USER,
  SIGNIN_SUCCESS,
  SIGNIN_FAILURE,
  IS_LOADING,
  SET_PROFILE,
  LOGOUT_USER,
  GET_PROFILE_ERROR
} from '../actions/types';

const initialState = {
  loading: false,
  user: {},
  profile: {},
  errors: {},
  isAuthenticated: false
};

export default (state = initialState, action) => {
  switch (action.type) {
    case IS_LOADING:
      return {
        ...state,
        loading: action.payload,
        errors: {}
      };
    case SET_CURRENT_USER:
      return {
        ...state,
        isAuthenticated: true,
        user: action.payload,
        errors: {},
        loading: false
      };
    case SIGNIN_FAILURE:
      return {
        ...state,
        errors: action.payload,
        loading: false
      };
    case SIGNIN_SUCCESS:
      return {
        ...state,
        errors: {},
        loading: false
      };
    case SET_PROFILE:
      return {
        ...state,
        profile: action.payload
      };
    case LOGOUT_USER:
      return {
        ...state,
        isAuthenticated: false,
        user: {},
        profile: {},
        loading: false
      };
    case GET_PROFILE_ERROR:
      return {
        ...state,
        errors: action.payload
      };
    default:
      return state;
  }
};
