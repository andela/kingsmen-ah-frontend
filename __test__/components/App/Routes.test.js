import React from 'react';
import Routes from '@components/App/Routes';
import { createStore, combineReducers } from 'redux';
import { BrowserRouter, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';
import auth from '@reducers/auth';

describe('Routes', () => {
    let store, history;

    beforeEach(() => {
        store = createStore(combineReducers({
            auth
        }));
        history = {
          push: jest.fn()
        }
    });

    it('should render index page', () => {
        const comp = (
          <Provider store={store}>
            <BrowserRouter>
              <Switch>
                <Routes dispatch={jest.fn()} />
              </Switch>
            </BrowserRouter>
          </Provider>
);

        const wrapper = mount(comp);
  
        expect(wrapper.find('Home')).toBeTruthy();
    });

    it('should render not found page', () => {
        const comp = (
          <Provider store={store}>
            <BrowserRouter>
              <Switch>
                <Routes dispatch={jest.fn()} />
              </Switch>
            </BrowserRouter>
          </Provider>
);
        const wrapper = mount(comp);
        history.push('/not-found');

        expect(wrapper.find('NotFound')).toBeTruthy();
    });
  

    it('should render the Routes component correctly', () => {
        const shallowWrapper = shallow(<Routes />);

        expect(toJson(shallowWrapper)).toMatchSnapshot();
    });
});