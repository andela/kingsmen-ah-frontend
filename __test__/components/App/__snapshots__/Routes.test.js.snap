// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes should render the Routes component correctly 1`] = `
<BrowserRouter>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/login"
    />
    <Route
      component={[Function]}
      path="/"
    />
  </Switch>
</BrowserRouter>
`;
